### PR TITLE
Fix SLE11-SP4 Incidents

### PR DESCRIPTION
**Pull request by dzedro**
- Related ticket: https://progress.opensuse.org/issues/125606
- Verification run:
https://openqa.suse.de/tests/overview?distri=sle&version=11-SP4&build=%3Atest%3A29128&groupid=485
https://openqa.suse.de/tests/overview?distri=sle&version=11-SP4&build=%3Atest%3A29130&groupid=485
https://dzedro.suse.cz/tests/1358#step/patch_and_reboot/19
https://dzedro.suse.cz/tests/1343 kernel_update 29128
https://dzedro.suse.cz/tests/1342 kernel_update 29130

### CODE REVIEW

File: lib/qam.pm [hunk 1]
```diff
@@ -73,31 +73,42 @@ sub is_patch_needed {
     }
 }
 
+sub add_repo_if_not_present {
+    my ($url, $name) = @_;
+    my $gpg = get_var('BUILD') =~ m/^MR:/ ? "-G" : "";
+    my $system_repos = script_output('zypper lr -u');
+    unless ($system_repos =~ /$url/) {
```
Comment (mdoucha) [inline lib/qam.pm:80]: I think this regex check could easily break due to unescaped special characters in `$url`. It'd be safer to call `$repolist = zypper_repos('-u')` and then `grep { $_->{uri} eq $url } @$repolist`

File: lib/qam.pm [hunk 2]
```diff
@@ -73,31 +73,42 @@ sub is_patch_needed {
...
+        add_repo_if_not_present("http://dist.suse.de/ibs/SUSE/Updates/SLE-SERVER/12-SP2-LTSS-ERICSSON/$arch/update/", '12-SP2-LTSS-ERICSSON-Updates');
+    }
+    elsif (is_sle('=12-SP3')) {
+        add_repo_if_not_present("http://dist.suse.de/ibs/SUSE/Updates/SLE-SERVER/12-SP3-LTSS-TERADATA/$arch/update/", '12-SP3-LTSS-TERADATA-Updates');
+    }
+    elsif (is_sle('=15-SP3')) {
+        add_repo_if_not_present("http://dist.suse.de/ibs/SUSE/Updates/SLE-Product-SLES/15-SP3-ERICSSON/$arch/update/", '15-SP3-ERICSSON-Updates');
+    }
+    elsif (is_sle('=15-SP4')) {
+        add_repo_if_not_present("http://dist.suse.de/ibs/SUSE/Updates/SLE-Product-SLES/15-SP4-ERICSSON/$arch/update/", '15-SP4-ERICSSON-Updates');
```
Comment (mdoucha) [inline lib/qam.pm:99]: When there are so many repos, it'd be cleaner to define an array and loop over it:

    my @repo_list = (
        {cond => '=12-SP2', name => '12-SP2-LTSS-ERICSSON-Updates', uri => '...'},
        {...}
    );
    
    for my $repo in (@repo_list) {
        add_repo_if_not_present($repo->{uri}, $repo->{name}) if is_sle($repo->{cond});
    }

**Review by mdoucha — CHANGES_REQUESTED** [review]

File: lib/qam.pm [hunk 3]
```diff
@@ -73,31 +73,42 @@ sub is_patch_needed {
...
+    my ($url, $name) = @_;
+    my $gpg = get_var('BUILD') =~ m/^MR:/ ? "-G" : "";
+    my $system_repos = script_output('zypper lr -u');
+    unless ($system_repos =~ /$url/) {
+        zypper_call("--no-gpg-checks ar -f $gpg -n '$name' $url '$name'");
+    }
+}
+
+# https://progress.opensuse.org/issues/90522
+sub add_extra_customer_repositories {
```
Comment (foursixnine) [inline lib/qam.pm:86]: I thought this already existed 🧠 ... deja vu!

**Comment by foursixnine** [comment]
waiting on @mdoucha's comments

File: lib/qam.pm (hunk 1 above)
Comment (dzedro) [reply lib/qam.pm]: Done, I don't know, but I don't think there should be any special characters in this urls, but now we will never find out. :wink:

File: lib/qam.pm (hunk 2 above)
Comment (dzedro) [reply lib/qam.pm]: I had feeling to make it better, this is great. Thank you

**Comment by mdoucha** [comment]
Code looks good but please add one more verification run.

**Comment by dzedro** [comment]
> Code looks good but please add one more verification run.

https://dzedro.suse.cz/tests/1379#step/patch_and_reboot/19
https://dzedro.suse.cz/tests/1375 kernel_update 29128
https://dzedro.suse.cz/tests/1376 kernel_update 29130

**Review by mdoucha — APPROVED** [review]
LGTM

**Review by foursixnine — APPROVED** [review]

**Pull request merged**